### ISSUE TITLE
MD013 sentence-per-line: a sentence ending inside emphasis is joined onto the next, producing an unfixable line-length violation

### ISSUE THREAD

**Issue by no-simpler** [rvben/rumdl#811]
**rumdl version:** 0.2.54 (also reproduced on every release from 0.2.45 onward)
**Installation method:** Pre-built binary

### Bug description

In `reflow-mode = "sentence-per-line"`, a sentence whose terminator sits inside an emphasis span is not recognised as finished. The span is joined onto the sentence that follows it, and the joined line exceeds `line-length` — so `rumdl fmt` turns conforming input into an `MD013` violation that `rumdl fmt` cannot then fix.

This is the same shape as #748 and #760: `fmt` produces output that `check` rejects.

### Steps to reproduce

`.rumdl.toml`:

```toml
[MD013]
line-length = 120
reflow = true
reflow-mode = "sentence-per-line"
```

`test.md` — three prose lines, none of them over `line-length`:

```markdown
# T

**An enum column may not be an atom a keyset scan resumes from.**
`ORDER BY` would follow the declaration index while the keyset predicate followed the value.
The two disagree, and the page skips rows.
```

```console
$ rumdl check test.md
test.md:3:1: [MD013] Paragraph should have one sentence per line (found 2 sentences across 3 lines) [*]

$ rumdl fmt test.md
test.md:3:1: [MD013] Paragraph should have one sentence per line (found 2 sentences across 3 lines) [fixed]

$ rumdl check test.md
test.md:3:121: [MD013] Line length 158 exceeds 120 characters

$ rumdl fmt test.md
test.md:3:121: [MD013] Line length 158 exceeds 120 characters
```

### Expected behavior

The input is already one sentence per line, so `check` reports nothing and `fmt` leaves it alone. This is what 0.2.44 does.

### Actual behavior

`check` counts 2 sentences where there are 3. `fmt` joins lines 3 and 4 into a single 158-character line, then reports it as a line-length violation with no fix available — `rumdl rule MD013` advertises `Fix: Fix is always available`. The rewrite is not idempotent and leaves the file worse than it found it.

The blast radius is not limited to one file. On a 49-file corpus that 0.2.43 reports clean, 0.2.54 reports 13 findings, rewrites 8 files, and leaves 5 unfixable `MD013` line-length findings behind. Every one of the 13 is a bold lead-in ending `.**` followed by a line opening with a code span or a lowercase word, which is an ordinary Markdown idiom rather than a corner case.

### Regression range

Bisected over the released binaries: 0.2.44 leaves the reproducer's longest line at 92 characters, and 0.2.45 and every release since produce 158.

The change is `207823b664ea5ca34c2c7c1ec5b2335366ea5b17`, "fix(md013): keep every word when a sentence boundary falls inside emphasis". It removed `handle_emphasis_sentence_split`, which tested the emphasis **content** for a terminator, and folded emphasis into the ordinary text path, which tests the **rendered span**. The rendered form of `**… resumes from.**` ends in `*`, not `.`, so `ends_with_sentence_punct` at `src/utils/text_reflow.rs:905` stopped firing for the single-sentence fast path in `reflow_elements_sentence_per_line`.

The plain-text analogue still works, which is why this only shows up with emphasis: for `Plain lead sentence.` followed by a code span the parser emits a text element ending in `.`, the fast path fires, and the line is emitted. An emphasis span is the one sentence-ending shape that fails the test, purely because of its closing marker.

Italic and strikethrough are affected identically:

```markdown
*Ital lead.*
`Code` next.
Tail here.
```

### Configuration

As above. `require-sentence-capital` is left at its default `true`; setting it to `false` masks the symptom but trades it for split-after-abbreviation false positives, so it is not a workaround.

I have a fix and tests ready and will open a PR against this issue.

**Comment by gotgenes** [comment]
Another data point on the blast radius, from a docs-heavy monorepo using the same `reflow-mode = "sentence-per-line"` config.

Bisecting released binaries against our corpus matches your range exactly:

| Version | Findings |
| --- | --- |
| 0.2.24 | 0 (clean, 1086 files) |
| 0.2.44 | 7 |
| 0.2.45 | regression begins |
| 0.2.62 | 767 across 333 files |

`rumdl fmt` rewrites all 767, joining lines that were already one sentence each.

We've pinned to 0.2.24. Examples of the shape in the wild, all bold lead-ins ending `.**` followed by a line opening with a code span:
https://github.com/gotgenes/pi-packages/blob/main/packages/pi-permission-system/docs/retro/0610-cross-session-decision-correlation.md

Thanks for the detailed report — the bisect saved me a lot of time.

**Comment by in-style** [comment]
Thanks for the second corpus — the shape matches mine exactly, including the bold lead-in ending `.**` followed by a line opening with a code span.

I ran the file you linked through three builds, with `reflow = true`, `reflow-mode = "sentence-per-line"`, `line-length = 120`:

| build | findings before `fmt` | after `fmt` | file after `fmt` |
| --- | --- | --- | --- |
| 0.2.43 | 61 | 61 | byte-identical to the original |
| 0.2.62 | 58 | 61 | four bold-label lines joined |
| #812 | 61 | 61 | byte-identical to the original |

The 61 are pre-existing long lines, since the file is not written one sentence per line — they sit there under every build and are not what this is about. The interesting column is the last one. 0.2.62 counts three fewer sentences than there are, joins four lines onto their lead-ins, and turns those into three fresh `Line length … exceeds` findings it then cannot fix. #812 leaves the file exactly as 0.2.43 does.

One practical note from your bisect: 0.2.43 is clean on my corpus, so the pin can probably move up from 0.2.24 and pick up eighteen releases of other fixes while this is open. Your table has 0.2.44 at 7 findings, which suggests something else moved between 0.2.43 and 0.2.44 in your tree. If those 7 are a different shape, I would be glad to see one — it may be a separate regression rather than this one.

Still live at the latest release, on a 95-file corpus written under this mode:

| build | findings | files `fmt` rewrites | findings left after `fmt` |
| --- | --- | --- | --- |
| 0.2.43 | 0 | 0 | 0 |
| 0.2.62 | 16 | 10 | 8, all unfixable length findings |
| #812 | 3 | 3 | 2 |

The 2 left are genuinely over-long single sentences that want a human, not lines `fmt` created, and a second `fmt` changes nothing.

#812 has just been rebased onto `main` and takes the maintainer's second round of review.

**Comment by no-simpler** [comment]
Thanks for the second corpus — the shape matches mine exactly, including the bold lead-in ending `.**` followed by a line opening with a code span.

I ran the file you linked through three builds, with `reflow = true`, `reflow-mode = "sentence-per-line"`, `line-length = 120`:

| build | findings before `fmt` | after `fmt` | file after `fmt` |
| --- | --- | --- | --- |
| 0.2.43 | 61 | 61 | byte-identical to the original |
| 0.2.62 | 58 | 61 | four bold-label lines joined |
| #812 | 61 | 61 | byte-identical to the original |

The 61 are pre-existing long lines, since the file is not written one sentence per line — they sit there under every build and are not what this is about. The interesting column is the last one. 0.2.62 counts three fewer sentences than there are, joins four lines onto their lead-ins, and turns those into three fresh `Line length … exceeds` findings it then cannot fix. #812 leaves the file exactly as 0.2.43 does.

One practical note from your bisect: 0.2.43 is clean on my corpus, so the pin can probably move up from 0.2.24 and pick up eighteen releases of other fixes while this is open. Your table has 0.2.44 at 7 findings, which suggests something else moved between 0.2.43 and 0.2.44 in your tree. If those 7 are a different shape, I would be glad to see one — it may be a separate regression rather than this one.

Still live at the latest release, on a 95-file corpus written under this mode:

| build | findings | files `fmt` rewrites | findings left after `fmt` |
| --- | --- | --- | --- |
| 0.2.43 | 0 | 0 | 0 |
| 0.2.62 | 16 | 10 | 8, all unfixable length findings |
| #812 | 3 | 3 | 2 |

The 2 left are genuinely over-long single sentences that want a human, not lines `fmt` created, and a second `fmt` changes nothing.

#812 has just been rebased onto `main` and takes the maintainer's second round of review.

**Comment by rvben** [comment]
Thanks @no-simpler, and apologies this one sat for a while.

PR #812 fixed it and v0.2.64 contains it: your reproducer is left byte-identical instead of being joined into a 158-character line. #816 is a different defect in the same mode and stays open on its own.
